geomstats/geometry/product_manifold.py

```diff
@@ -14,101 +14,288 @@
+                    shape = (len(self.factors), *self.factors[0].shape)
+                else:
+                    raise ValueError(
+                        "A default_point_type of \'matrix\' can only be used if all "
+                        "manifolds have vector type."
+                    )
+            else:
+                raise ValueError(
+                    "A default_point_type of \'matrix\' can only be used if all "
+                    "manifolds have the same shape."
+                )
+
+        if "extrinsic" in self.factor_default_coords_types:
+            default_coords_type = "extrinsic"
+        else:
+            default_coords_type = "intrinsic"
+
         if metrics is None:
-            metrics = [manifold.metric for manifold in manifolds]
+            metric_scales = None
+            metrics = [manifold.metric for manifold in factors]
+        elif gs.all([is_positive(metric) for metric in metrics]):
+            metric_scales = metrics
+            metrics = [manifold.metric for manifold in factors]
+        else:
+            metric_scales = None
+
         kwargs.setdefault(
             "metric",
             ProductRiemannianMetric(
-                metrics, n_jobs=n_jobs, default_point_type=default_point_type
-            ),
+                metrics, default_point_type=default_point_type, scales=metric_scales),
         )
-        dim = sum(self.dims)
-
-        if default_point_type == "vector":
-            shape = (sum([m.shape[0] for m in manifolds]),)
-        else:
-            shape = (len(manifolds), *manifolds[0].shape)
 
         super().__init__(
             dim=dim,
             shape=shape,
+            default_coords_type=default_coords_type,
             **kwargs,
         )
-        self.manifolds = manifolds
-        self.n_jobs = n_jobs
+
+        if self.default_coords_type == "extrinsic":
+            factor_embedding_spaces = [
+                manifold.embedding_space if manifold.default_coords_type == "extrinsic"
+                else manifold
+                for manifold in factors
+            ]
+            self.embedding_space = ProductManifold(factor_embedding_spaces)
+
+    @property
+    def factor_dims(self):
+        """List containing the dimension of each factor."""
+        return [factor.dim for factor in self.factors]
+
+    @property
+    def factor_shapes(self):
+        """List containing the shape of each factor."""
+        return [factor.shape for factor in self.factors]
+
+    @property
+    def factor_default_coords_types(self):
+        """List containing the default_coords_type of each factor."""
+        return [factor.default_coords_type for factor in self.factors]
 
     @staticmethod
     def _get_method(manifold, method_name, metric_args):
         return getattr(manifold, method_name)(**metric_args)
 
-    def _iterate_over_manifolds(self, func, args, intrinsic=False):
+    def _validate_args_for_iteration(self, args):
```

Comment:
the name of this function is a bit misleading, as it does more than validation: `_validate_and_prepare_args_for_iteration` instead?